pkg/registry/blobs.go

```diff
@@ -76,6 +76,35 @@ type blobPutHandler interface {
 	Put(ctx context.Context, repo string, h v1.Hash, rc io.ReadCloser) error
 }
 
+// uploadHandler represents a minimal upload storage backend, capable of
+// appending streamed upload contents and finally returning them to be stored
+// in blob storage.
+type uploadHandler interface {
+	// StatUpload returns the current size of the given upload.
+	StatUpload(ctx context.Context, uploadID string) (int64, error)
+
+	// AppendUpload appends the contents of the ReadCloser to the current
+	// upload contents, and returns the new total size.
+	AppendUpload(ctx context.Context, uploadID string, rc io.ReadCloser) (int64, error)
+
+	// FinishUpload appends the contents of the ReadCloser to the curent
+	// upload contents, and returns the total contents and their size.
+	FinishUpload(ctx context.Context, uploadID string, rc io.ReadCloser) (io.ReadCloser, int64, error)
```

Comment:
These names bother me. I think I want `GetUpload(uploadID) (io.ReadCloser, error)`, which naive upload finalization will call after `Append`ing the new bits, and then `DeleteUpload(uploadID) error` before `PutBlob` with the full upload contents.

Implementations that support `FinalizeUpload` can just have `GetUpload` and `DeleteUpload` always return an error since that's what `FinalizeUpload` is expected to do.